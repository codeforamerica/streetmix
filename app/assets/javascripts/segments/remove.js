var TRACK_ACTION_REMOVE_SEGMENT = 'Remove segment';
var TRACK_LABEL_BUTTON = 'Button';

function _removeSegment(el, all) {
  if (all) {
    street.segments = [];
    _createDomFromData();
    _segmentsChanged();

    _infoBubble.hide();

    _statusMessage.show(msg('STATUS_ALL_SEGMENTS_DELETED'), true);
  } else if (el && el.parentNode) {
    _infoBubble.hide();
    _infoBubble.hideSegment();
    _switchSegmentElAway(el);
    _segmentsChanged();

    _statusMessage.show(msg('STATUS_SEGMENT_DELETED'), true);
  }

  /*if (street.segments.length) {
    _showWidthChartImmediately();
    _hideWidthChart();
  }*/
}

function _onRemoveButtonClick(event) {
  var el = event.target.segmentEl;

  if (el) {
    _removeSegment(el, event.shiftKey);

    Stmx.app.eventTracking.track(TRACK_CATEGORY_INTERACTION, TRACK_ACTION_REMOVE_SEGMENT,
        TRACK_LABEL_BUTTON, null, true);
  }

  // Prevent this “leaking” to a segment below
  event.preventDefault();
}
