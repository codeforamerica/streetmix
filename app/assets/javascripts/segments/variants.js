var VARIANT_SEPARATOR = '|';

var VARIANTS = {
  'direction': ['inbound', 'outbound'],
  'parking-lane-direction': ['inbound', 'outbound', 'sideways'],

  'tree-type': ['big', 'palm-tree'],

  'lamp-orientation': ['left', 'both', 'right'],
  'lamp-type': ['modern', 'traditional'],

  'bench-orientation': ['left', 'center', 'right'],
  'turn-lane-orientation': ['left', 'left-straight', 'straight', 'right-straight', 'right',  'both', 'shared'],

  'divider-type': ['median', 'striped-buffer', 'planting-strip',
                   'planter-box', 'bush', 'flowers', 'big-tree',
                   'palm-tree', 'bollard', 'dome'],

  'orientation': ['left', 'right'],

  'public-transit-asphalt': ['regular', 'colored'],
  'bike-asphalt': ['regular', 'colored'],

  'transit-shelter-elevation': ['street-level', 'light-rail'],
  'bike-rack-elevation': ['sidewalk-parallel', 'sidewalk', 'road'],

  'car-type': ['car', 'sharrow', 'truck'],
  'sidewalk-density': ['dense', 'normal', 'sparse', 'empty'],

  'parking-lane-orientation': ['left', 'right'],
  'wayfinding-type': ['large', 'medium', 'small']
};

function _getVariantArray(segmentType, variantString) {
  var variantArray = {};
  var variantSplit = variantString.split(VARIANT_SEPARATOR);

  for (var i in SEGMENT_INFO[segmentType].variants) {
    var variantName = SEGMENT_INFO[segmentType].variants[i];

    variantArray[variantName] = variantSplit[i];
  }

  return variantArray;
}

function _getVariantString(variant) {
  var string = '';
  for (var i in variant) {
    string += variant[i] + VARIANT_SEPARATOR;
  }

  string = string.substr(0, string.length - 1);
  return string;
}
