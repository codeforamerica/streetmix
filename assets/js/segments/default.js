var DEFAULT_SEGMENTS = {
  'false': [ // Right-hand traffic
    { type: 'sidewalk', variant: { 'sidewalk-density': 'dense' }, width: 6 },
    { type: 'sidewalk-tree', variant: { 'tree-type': 'big' }, width: 2 },
    { type: 'transit-shelter', variant: { 'orientation': 'left', 'transit-shelter-elevation': 'street-level' }, width: 9 },
    { type: 'sidewalk-lamp', variant: { 'lamp-orientation': 'right', 'lamp-type': 'modern' }, width: 2 },
    { type: 'bus-lane', variant: { 'direction': 'inbound', 'public-transit-asphalt': 'regular' }, width: 10 },
    { type: 'turn-lane', variant: { 'direction': 'inbound', 'turn-lane-orientation': 'right-straight'}, width: 10 },
    { type: 'divider', variant: { 'divider-type': 'flowers' }, width: 3 },
    { type: 'drive-lane', variant: { 'direction': 'outbound', 'car-type': 'truck' }, width: 10 },
    { type: 'bike-lane', variant: { 'direction': 'outbound', 'bike-asphalt': 'colored' }, width: 6 },
    { type: 'parking-lane', variant: { 'parking-lane-direction': 'outbound', 'parking-lane-orientation': 'right' }, width: 8 },
    { type: 'sidewalk-lamp', variant: { 'lamp-orientation': 'left', 'lamp-type': 'modern' }, width: 2 },
    { type: 'sidewalk-tree', variant: { 'tree-type': 'big' }, width: 2 },
    { type: 'sidewalk-wayfinding', variant: { 'wayfinding-type': 'medium' }, width: 2 },
    { type: 'sidewalk', variant: { 'sidewalk-density': 'normal' }, width: 6 },
    { type: 'sidewalk-bench', variant: { 'bench-orientation': 'right' }, width: 2}
  ],
  'true': [ // Left-hand traffic
    { type: 'sidewalk-bench', variant: { 'bench-orientation': 'left' }, width: 2},
    { type: 'sidewalk', variant: { 'sidewalk-density': 'normal' }, width: 6 },
    { type: 'sidewalk-wayfinding', variant: { 'wayfinding-type': 'medium' }, width: 2 },
    { type: 'sidewalk-tree', variant: { 'tree-type': 'big' }, width: 2 },
    { type: 'sidewalk-lamp', variant: { 'lamp-orientation': 'right', 'lamp-type': 'modern' }, width: 2 },
    { type: 'parking-lane', variant: { 'parking-lane-direction': 'outbound', 'parking-lane-orientation': 'left' }, width: 8 },
    { type: 'bike-lane', variant: { 'direction': 'outbound', 'bike-asphalt': 'colored' }, width: 6 },
    { type: 'drive-lane', variant: { 'direction': 'outbound', 'car-type': 'truck' }, width: 10 },
    { type: 'divider', variant: { 'divider-type': 'flowers' }, width: 3 },
    { type: 'turn-lane', variant: { 'direction': 'inbound', 'turn-lane-orientation': 'left-straight'}, width: 10 },
    { type: 'bus-lane', variant: { 'direction': 'inbound', 'public-transit-asphalt': 'regular' }, width: 10 },
    { type: 'sidewalk-lamp', variant: { 'lamp-orientation': 'left', 'lamp-type': 'modern' }, width: 2 },
    { type: 'transit-shelter', variant: { 'orientation': 'right', 'transit-shelter-elevation': 'street-level' }, width: 9 },
    { type: 'sidewalk-tree', variant: { 'tree-type': 'big' }, width: 2 },
    { type: 'sidewalk', variant: { 'sidewalk-density': 'dense' }, width: 6 }
  ]
};
