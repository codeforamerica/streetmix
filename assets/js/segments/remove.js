var TRACK_ACTION_REMOVE_SEGMENT = 'Remove segment'
var TRACK_LABEL_BUTTON = 'Button'

function _removeSegment (el, all) {
  if (all) {
    street.segments = []
    _createDomFromData()
    _segmentsChanged()

    _infoBubble.hide()

    showStatusMessage(msg('STATUS_ALL_SEGMENTS_DELETED'), true)
  } else if (el && el.parentNode) {
    _infoBubble.hide()
    _infoBubble.hideSegment()
    _switchSegmentElAway(el)
    _segmentsChanged()

    showStatusMessage(msg('STATUS_SEGMENT_DELETED'), true)
  }
}

function _onRemoveButtonClick (event) {
  var el = event.target.segmentEl

  if (el) {
    _removeSegment(el, event.shiftKey)

    trackEvent('Interaction', TRACK_ACTION_REMOVE_SEGMENT,
      TRACK_LABEL_BUTTON, null, true)
  }

  // Prevent this “leaking” to a segment below
  event.preventDefault()
}
