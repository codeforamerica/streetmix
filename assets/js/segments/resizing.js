var SHORT_DELAY = 100

var RESIZE_TYPE_INITIAL = 0
var RESIZE_TYPE_INCREMENT = 1
var RESIZE_TYPE_DRAGGING = 2
var RESIZE_TYPE_PRECISE_DRAGGING = 3
var RESIZE_TYPE_TYPING = 4

var MIN_SEGMENT_WIDTH = 1
var MAX_SEGMENT_WIDTH = 400

var TOUCH_CONTROLS_FADEOUT_TIME = 3000
var TOUCH_CONTROLS_FADEOUT_DELAY = 3000

var NORMALIZE_PRECISION = 5

var segmentWidthResolution
var segmentWidthClickIncrement
var segmentWidthDraggingResolution

var suppressMouseEnter = false

function _resizeSegment (el, resizeType, width, updateEdit, palette, initial) {
  if (!palette) {
    var width =
    _normalizeSegmentWidth(width / TILE_SIZE, resizeType) * TILE_SIZE
  }

  document.body.classList.add('immediate-segment-resize')

  window.setTimeout(function () {
    document.body.classList.remove('immediate-segment-resize')
  }, SHORT_DELAY)

  var oldWidth = parseFloat(el.getAttribute('width') * TILE_SIZE)

  el.style.width = width + 'px'
  el.setAttribute('width', width / TILE_SIZE)

  var widthEl = el.querySelector('span.width')
  if (widthEl) {
    widthEl.innerHTML =
      _prettifyWidth(width / TILE_SIZE, { markup: true })
  }

  _setSegmentContents(el, el.getAttribute('type'),
    el.getAttribute('variant-string'), width, parseInt(el.getAttribute('rand-seed')), palette, false)

  if (updateEdit) {
    _infoBubble.updateWidthInContents(el, width / TILE_SIZE)
  }

  if (!initial) {
    _segmentsChanged()

    let segment = street.segments[parseInt(el.dataNo)]
    _infoBubble.updateWarningsInContents(segment)
  }
}

function _handleSegmentResizeCancel () {
  _resizeSegment(draggingResize.segmentEl, RESIZE_TYPE_INITIAL,
    draggingResize.originalWidth * TILE_SIZE, true, false)

  _handleSegmentResizeEnd()
}

function _handleSegmentResizeEnd (event) {
  ignoreStreetChanges = false

  _segmentsChanged()

  _changeDraggingType(DRAGGING_TYPE_NONE)

  var el = draggingResize.floatingEl
  removeElFromDOM(el)

  draggingResize.segmentEl.classList.add('immediate-show-drag-handles')

  _removeGuides(draggingResize.segmentEl)

  _infoBubble.considerSegmentEl = draggingResize.segmentEl
  _infoBubble.show(false)

  _scheduleControlsFadeout(draggingResize.segmentEl)

  suppressMouseEnter = true
  _infoBubble.considerShowing(event, draggingResize.segmentEl, INFO_BUBBLE_TYPE_SEGMENT)
  window.setTimeout(function () {
    suppressMouseEnter = false
  }, 50)

  if (draggingResize.width && (draggingResize.originalWidth != draggingResize.width)) {
    trackEvent('INTERACTION', 'CHANGE_WIDTH', 'DRAGGING', null, true)
  }
}

function _normalizeAllSegmentWidths () {
  for (var i in street.segments) {
    street.segments[i].width =
      _normalizeSegmentWidth(street.segments[i].width, RESIZE_TYPE_INITIAL)
  }
}

function _normalizeSegmentWidth (width, resizeType) {
  if (width < MIN_SEGMENT_WIDTH) {
    width = MIN_SEGMENT_WIDTH
  } else if (width > MAX_SEGMENT_WIDTH) {
    width = MAX_SEGMENT_WIDTH
  }

  switch (resizeType) {
    case RESIZE_TYPE_INITIAL:
    case RESIZE_TYPE_TYPING:
    case RESIZE_TYPE_INCREMENT:
    case RESIZE_TYPE_PRECISE_DRAGGING:
      var resolution = segmentWidthResolution
      break
    case RESIZE_TYPE_DRAGGING:
      var resolution = segmentWidthDraggingResolution
      break
  }

  width = Math.round(width / resolution) * resolution
  width = parseFloat(width.toFixed(NORMALIZE_PRECISION))

  return width
}

function _incrementSegmentWidth (segmentEl, add, precise) {
  var width = parseFloat(segmentEl.getAttribute('width'))

  if (precise) {
    var increment = segmentWidthResolution
  } else {
    var increment = segmentWidthClickIncrement
  }

  if (!add) {
    increment = -increment
  }
  width = _normalizeSegmentWidth(width + increment, RESIZE_TYPE_INCREMENT)

  _resizeSegment(segmentEl, RESIZE_TYPE_INCREMENT,
    width * TILE_SIZE, true, false)
}

function _applyWarningsToSegments () {
  for (var i in street.segments) {
    var segment = street.segments[i]

    if (segment.el) {
      if (segment.warnings[SEGMENT_WARNING_OUTSIDE] ||
        segment.warnings[SEGMENT_WARNING_WIDTH_TOO_SMALL] ||
        segment.warnings[SEGMENT_WARNING_WIDTH_TOO_LARGE]) {
        segment.el.classList.add('warning')
      } else {
        segment.el.classList.remove('warning')
      }

      if (segment.warnings[SEGMENT_WARNING_OUTSIDE]) {
        segment.el.classList.add('outside')
      } else {
        segment.el.classList.remove('outside')
      }
    }
  }
}

var controlsFadeoutDelayTimer = -1
var controlsFadeoutHideTimer = -1

function _scheduleControlsFadeout (el) {
  _infoBubble.considerShowing(null, el, INFO_BUBBLE_TYPE_SEGMENT)

  _resumeFadeoutControls()
}

function _resumeFadeoutControls () {
  if (!system.touch) {
    return
  }

  _cancelFadeoutControls()

  controlsFadeoutDelayTimer = window.setTimeout(_fadeoutControls, TOUCH_CONTROLS_FADEOUT_DELAY)
}

function _cancelFadeoutControls () {
  document.body.classList.remove('controls-fade-out')
  window.clearTimeout(controlsFadeoutDelayTimer)
  window.clearTimeout(controlsFadeoutHideTimer)
}

function _fadeoutControls () {
  document.body.classList.add('controls-fade-out')

  controlsFadeoutHideTimer = window.setTimeout(_hideControls, TOUCH_CONTROLS_FADEOUT_TIME)
}

function _hideControls () {
  document.body.classList.remove('controls-fade-out')
  if (_infoBubble.segmentEl) {
    _infoBubble.segmentEl.classList.remove('show-drag-handles')

    window.setTimeout(function () {
      _infoBubble.hide()
      _infoBubble.hideSegment(true)
    }, 0)
  }
}
