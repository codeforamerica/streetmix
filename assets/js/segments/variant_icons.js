var VARIANT_ICONS = {
  'trashcan': { x: 3, y: 1 },

  'orientation|left': { x: 0, y: 0, title: 'Left' },
  'orientation|right': { x: 1, y: 0, title: 'Right' },
  'turn-lane-orientation|left': { x: 7, y: 1, title: 'Left' },
  'turn-lane-orientation|left-straight': { x: 8, y: 1, title: 'Left and straight' },
  'turn-lane-orientation|straight': { x: 6, y: 1, title: 'Straight' },
  'turn-lane-orientation|right-straight': { x: 9, y: 1, title: 'Right and straight' },
  'turn-lane-orientation|right': { x: 4, y: 1, title: 'Right' },
  'turn-lane-orientation|both': { x: 5, y: 1, title: 'Both' },
  'turn-lane-orientation|shared': { x: 7, y: 0, title: 'Shared' },
  'bench-orientation|left': { x: 4, y: 0, title: 'Left' },
  'bench-orientation|right': { x: 5, y: 0, title: 'Right' },
  'bench-orientation|center': { x: 6, y: 0, title: 'Center' },
  'parking-lane-orientation|left': { x: 0, y: 0, title: 'Left' },
  'parking-lane-orientation|right': { x: 1, y: 0, title: 'Right' },
  'parking-lane-direction|inbound': { x: 2, y: 0, title: 'Inbound' },
  'parking-lane-direction|outbound': { x: 3, y: 0, title: 'Outbound' },
  'parking-lane-direction|sideways': { x: 6, y: 0, title: 'Perpendicular' },
  'direction|inbound': { x: 2, y: 0, title: 'Inbound' },
  'direction|outbound': { x: 3, y: 0, title: 'Outbound' },
  'sidewalk-density|empty': { x: -1, y: 1, title: 'Empty' },
  'sidewalk-density|sparse': { x: 0, y: 1, title: 'Sparse' },
  'sidewalk-density|normal': { x: 1, y: 1, title: 'Normal' },
  'sidewalk-density|dense': { x: 2, y: 1, title: 'Dense' },
  'lamp-orientation|left': { x: 4, y: 0, title: 'Left' },
  'lamp-orientation|both': { x: 6, y: 0, title: 'Both' },
  'lamp-orientation|right': { x: 5, y: 0, title: 'Right' },
  'lamp-type|traditional': { x: 4, y: 2, title: 'Traditional' },
  'lamp-type|modern': { x: 3, y: 2, title: 'Modern' },
  'car-type|car': { x: 0, y: 3, title: 'Car' },
  'car-type|sharrow': { x: 5, y: 3, title: 'Sharrow' },
  'car-type|truck': { x: 1, y: 3, title: 'Truck' },
  'public-transit-asphalt|regular': { x: 2, y: 3, title: '?' },
  'public-transit-asphalt|colored': { x: 3, y: 3, title: '?' },
  'bike-asphalt|regular': { x: 2, y: 3, title: '?' },
  'bike-asphalt|colored': { x: 4, y: 3, title: '?' },
  'tree-type|big': { x: 1, y: 4, title: 'Tree' },
  'tree-type|palm-tree': { x: 0, y: 4, title: 'Palm tree' },
  'divider-type|median': { x: 7, y: 4, title: 'Median' },
  'divider-type|striped-buffer': { x: 8, y: 4, title: 'Striped buffer' },
  'divider-type|planting-strip': { x: 2, y: 4, title: 'Planting strip' },
  'divider-type|bush': { x: 4, y: 4, title: 'Planting strip with a bush' },
  'divider-type|flowers': { x: 5, y: 4, title: 'Planting strip with flowers' },
  'divider-type|big-tree': { x: 1, y: 4, title: 'Planting strip with a tree' },
  'divider-type|palm-tree': { x: 0, y: 4, title: 'Planting strip with a palm tree' },
  'divider-type|bollard': { x: 6, y: 4, title: 'Bollard' },
  'divider-type|dome': { x: 6, y: 3, title: 'Traffic exclusion dome' },
  'transit-shelter-elevation|street-level': { x: 5, y: 2, title: 'Street level' },
  'transit-shelter-elevation|light-rail': { x: 6, y: 2, title: 'Light rail platform' },
  'bike-rack-elevation|sidewalk-parallel': { x: 6, y: 0, title: 'Parallel parking, sidewalk level' },
  'bike-rack-elevation|sidewalk': { x: 6, y: 2, title: 'Perpendicular parking, sidewalk level' },
  'bike-rack-elevation|road': { x: 5, y: 2, title: 'Perpendicular parking, road level' },
  'building|waterfront': { x: 9, y: 4, title: 'Waterfront' },
  'building|grass': { x: 2, y: 4, title: 'Grass' },
  'building|fence': { x: 3, y: 4, title: 'Empty lot' },
  'building|parking-lot': { x: 0, y: 3, title: 'Parking lot' },
  'building|residential': { x: 7, y: 3, title: 'Home' },
  'building|narrow': { x: 7, y: 2, title: 'Narrow building' },
  'building|wide': { x: 8, y: 2, title: 'Wide building' },
  'wayfinding-type|large': { x: 8, y: 3, title: 'Large' },
  'wayfinding-type|medium': { x: 9, y: 3, title: 'Medium' },
  'wayfinding-type|small': { x: 10, y: 3, title: 'Small' }
};
