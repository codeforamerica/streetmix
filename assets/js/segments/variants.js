var VARIANT_SEPARATOR = '|';

var VARIANTS = {
  'direction': ['inbound', 'outbound'],
  'parking-lane-direction': ['inbound', 'outbound', 'sideways'],

  'tree-type': ['big', 'palm-tree'],

  'lamp-orientation': ['left', 'both', 'right'],
  'lamp-type': ['modern', 'traditional'],

  'bench-orientation': ['left', 'center', 'right'],
  'turn-lane-orientation': ['left', 'left-straight', 'straight', 'right-straight', 'right',  'both', 'shared'],

  'divider-type': ['median', 'striped-buffer', 'planting-strip',
                   'bush', 'flowers', 'big-tree',
                   'palm-tree', 'bollard', 'dome'],

  'orientation': ['left', 'right'],

  'public-transit-asphalt': ['regular', 'colored'],
  'bike-asphalt': ['regular', 'colored'],

  'transit-shelter-elevation': ['street-level', 'light-rail'],
  'bike-rack-elevation': ['sidewalk-parallel', 'sidewalk', 'road'],

  'car-type': ['car', 'sharrow', 'truck'],
  'sidewalk-density': ['dense', 'normal', 'sparse', 'empty'],

  'parking-lane-orientation': ['left', 'right'],
  'wayfinding-type': ['large', 'medium', 'small']
};
