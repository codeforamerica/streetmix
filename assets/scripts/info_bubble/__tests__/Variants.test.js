/* eslint-env jest */
import React from 'react'
import { Variants } from '../Variants'
import { shallow } from 'enzyme'
import { IntlProvider } from 'react-intl'

import { getSegmentInfo } from '../../segments/info'
import {
  INFO_BUBBLE_TYPE_SEGMENT,
  INFO_BUBBLE_TYPE_LEFT_BUILDING,
  INFO_BUBBLE_TYPE_RIGHT_BUILDING
} from '../constants'

jest.mock('../../streets/data_model', () => {})
jest.mock('../../segments/view')
jest.mock('../../segments/info')
jest.mock('../../segments/variant_icons.json', () => require('../../segments/__mocks__/variant_icons.json'))

describe('Variants', () => {
  const intlProvider = new IntlProvider({ locale: 'en' }, {})
  const { intl } = intlProvider.getChildContext()

  // Mock required props for component
  const requiredProps = {
    changeSegmentVariant: jest.fn(),
    setBuildingVariant: jest.fn(),
    flags: {},
    intl: intl
  }

  it('does not render component', () => {
    const wrapper = shallow(<Variants {...requiredProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  describe('INFO_BUBBLE_TYPE_SEGMENT', () => {
    let segment

    beforeEach(() => {
      segment = { variants: ['direction', 'public-transit-asphalt'] } // coming from info.json
      getSegmentInfo.mockImplementation(() => segment)
    })

    const type = INFO_BUBBLE_TYPE_SEGMENT
    const streetSegment = { variantString: 'inbound|regular', segmentType: 'streetcar' }

    it('renders segment type correctly', () => {
      const wrapper = shallow(<Variants {...requiredProps} type={type} position={0} variant={streetSegment.variantString} segmentType={streetSegment.segmentType} />)
      expect(wrapper).toMatchSnapshot()
    })

    it('onClick calls changeSegmentVariant', () => {
      const changeSegmentVariant = jest.fn()
      const wrapper = shallow(<Variants {...requiredProps} changeSegmentVariant={changeSegmentVariant} type={type} position={0} variant={streetSegment.variantString} segmentType={streetSegment.segmentType} />)
      wrapper.find('button').first().simulate('click')
      expect(changeSegmentVariant).toHaveBeenCalledTimes(1)
      expect(changeSegmentVariant).toHaveBeenCalledWith(0, 'direction', 'inbound')
    })
  })

  describe('INFO_BUBBLE_TYPE_LEFT_BUILDING', () => {
    const type = INFO_BUBBLE_TYPE_LEFT_BUILDING

    it('onClick calls setBuildingVariant', () => {
      const setBuildingVariant = jest.fn()
      const wrapper = shallow(<Variants {...requiredProps} setBuildingVariant={setBuildingVariant} type={type} position={'left'} variant={'residential'} />)
      wrapper.find('button').first().simulate('click')
      expect(setBuildingVariant).toHaveBeenCalledTimes(1)
      expect(setBuildingVariant).toHaveBeenCalledWith('left', 'waterfront')
    })

    it('renders left building correctly', () => {
      const wrapper = shallow(<Variants {...requiredProps} type={type} position={'left'} variant={'residential'} />)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('INFO_BUBBLE_TYPE_RIGHT_BUILDING', () => {
    const type = INFO_BUBBLE_TYPE_RIGHT_BUILDING

    it('onClick calls setBuildingVariant', () => {
      const setBuildingVariant = jest.fn()
      const wrapper = shallow(<Variants {...requiredProps} setBuildingVariant={setBuildingVariant} type={type} position={'right'} variant={'residential'} />)
      wrapper.find('button').first().simulate('click')
      expect(setBuildingVariant).toHaveBeenCalledTimes(1)
      expect(setBuildingVariant).toHaveBeenCalledWith('right', 'waterfront')
    })

    it('renders left building correctly', () => {
      const wrapper = shallow(<Variants {...requiredProps} type={type} position={'right'} variant={'residential'} />)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('flag', () => {
    let segment

    beforeEach(() => {
      segment = { variants: ['building-z', 'waterfront'] } // coming from info.json
      getSegmentInfo.mockImplementation(() => segment)
    })

    const type = INFO_BUBBLE_TYPE_SEGMENT
    const streetSegment = { variantString: 'inbound|regular', segmentType: 'streetcar' }
    const flags = { 'flag': { value: 't' } }

    it('renders with flag correctly', () => {
      const wrapper = shallow(<Variants {...requiredProps} type={type} position={0} variant={streetSegment.variantString} segmentType={streetSegment.segmentType} flags={flags} />)
      expect(wrapper).toMatchSnapshot()
    })
  })
})
