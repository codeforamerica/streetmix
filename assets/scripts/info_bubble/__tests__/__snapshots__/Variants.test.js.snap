// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Variants INFO_BUBBLE_TYPE_LEFT_BUILDING renders left building correctly 1`] = `
<div
  className="variants"
>
  <button
    disabled={false}
    key="building.waterfront"
    onClick={[Function]}
    title="Waterfront"
  >
    <svg
      className="icon"
      style={
        Object {
          "fill": "#ff0",
        }
      }
      xmlns="http://www.w3.org/1999/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <use
        xlinkHref="#icon-waterfront"
      />
    </svg>
  </button>
</div>
`;

exports[`Variants INFO_BUBBLE_TYPE_RIGHT_BUILDING renders left building correctly 1`] = `
<div
  className="variants"
>
  <button
    disabled={false}
    key="building.waterfront"
    onClick={[Function]}
    title="Waterfront"
  >
    <svg
      className="icon"
      style={
        Object {
          "fill": "#ff0",
        }
      }
      xmlns="http://www.w3.org/1999/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <use
        xlinkHref="#icon-waterfront"
      />
    </svg>
  </button>
</div>
`;

exports[`Variants INFO_BUBBLE_TYPE_SEGMENT renders segment type correctly 1`] = `
<div
  className="variants"
>
  <button
    disabled={true}
    key="direction.inbound"
    onClick={[Function]}
    title="Inbound"
  >
    <svg
      className="icon"
      style={null}
      xmlns="http://www.w3.org/1999/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <use
        xlinkHref="#icon-direction-inbound"
      />
    </svg>
  </button>
  <button
    disabled={false}
    key="direction.outbound"
    onClick={[Function]}
    title="Outbound"
  >
    <svg
      className="icon"
      style={null}
      xmlns="http://www.w3.org/1999/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <use
        xlinkHref="#icon-direction-outbound"
      />
    </svg>
  </button>
  <button
    disabled={false}
    key="direction.empty"
    onClick={[Function]}
    title="No one"
  >
    <svg
      className="icon"
      style={null}
      xmlns="http://www.w3.org/1999/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <use
        xlinkHref="#icon-"
      />
    </svg>
  </button>
  <hr
    key="public-transit-asphalt"
  />
</div>
`;

exports[`Variants does not render component 1`] = `""`;

exports[`Variants flag renders with flag correctly 1`] = `
<div
  className="variants"
>
  <button
    disabled={false}
    key="building-z.waterfront"
    onClick={[Function]}
    title="Waterfront"
  >
    <svg
      className="icon"
      style={null}
      xmlns="http://www.w3.org/1999/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <use
        xlinkHref="#icon-waterfront"
      />
    </svg>
  </button>
  <hr
    key="waterfront"
  />
</div>
`;
